packages/tools/src/utilities/segmentation/segmentContourAction.ts

```diff
@@ -0,0 +1,161 @@
+import { getEnabledElement, type Types } from '@cornerstonejs/core';
+
+import type { Annotation } from '../../types/AnnotationTypes';
+import * as segmentation from '../../stateManagement/segmentation';
+import {
+  state as annotationState,
+  config as annotationConfig,
+} from '../../stateManagement/annotation';
+import { jumpToSlice } from '../viewport';
+import contourAndFindLargestBidirectional from './contourAndFindLargestBidirectional';
+import createBidirectionalToolData from './createBidirectionalToolData';
+import BidirectionalTool from '../../tools/annotation/BidirectionalTool';
+
+export type Segment = {
+  segmentationId: string;
+  segmentIndex: number;
+  label: string;
+
+  style?: any;
+  containedSegmentIndices?: (number) => boolean;
+};
+
+export type SegmentContourActionConfiguration = {
+  getSegment?: (
+    enabledElement: Types.IEnabledElement,
+    configuration: SegmentContourActionConfiguration
+  ) => Segment;
+
+  /**
+   * Optional map for data about each segment
+   */
+  segmentationId?: string;
+  segmentData?: Map<number, Segment>;
+  toolGroupId?: string;
+};
+
+export default function segmentContourAction(
+  element: HTMLDivElement,
+  configuration: SegmentContourActionConfiguration
+) {
+  const enabledElement = getEnabledElement(element);
+  const segment = (configuration.getSegment || defaultGetSegment)(
+    enabledElement,
+    configuration
+  );
+  if (!segment) {
+    return;
+  }
+  const FrameOfReferenceUID = enabledElement.viewport.getFrameOfReferenceUID();
+  const segmentationsList = segmentation.state.getSegmentations();
+  const { segmentIndex, segmentationId } = segment;
+  const bidirectionals = annotationState.getAnnotations(
+    this.toolName || BidirectionalTool.toolName,
+    FrameOfReferenceUID
+  );
+  let hasExistingActiveSegment = false;
+  const existingSegments = bidirectionals.filter((it) => {
+    const { segment } = it.data;
+    if (!segment) {
+      return;
+    }
+    if (
+      segment.segmentationId === segmentationId &&
+      segment.segmentIndex === segmentIndex
+    ) {
+      hasExistingActiveSegment = true;
+      it.data.segment = segment;
+    }
+    return !!segment;
+  });
+  if (!hasExistingActiveSegment) {
+    // Just create a dummy annotation object containing just enough information
+    // to create a real one.
+    existingSegments.push({
+      data: { segment },
+    } as unknown as Annotation);
+  }
+
+  let newBidirectional;
+  existingSegments.forEach((annotation) => {
+    const segments = [];
+    const { segment: updateSegment } = annotation.data;
+    const { segmentIndex, segmentationId } = updateSegment;
+    segments[segmentIndex] = updateSegment;
+    annotationState.removeAnnotation(annotation.annotationUID);
+    const bidirectionalData = contourAndFindLargestBidirectional({
+      ...segmentationsList.find(
+        (segmentation) => segmentation.segmentationId === segmentationId
+      ),
```

Comment:
you don't need the ...

Reply:
How would it work otherwise?  I need a new object since I don't want to modify the original one, and it has to be the extension of the found one?